### ISSUE TITLE
Allow kubectl exec to proxy through docker exec

### ISSUE THREAD

**Issue by bprashanth** [GoogleCloudPlatform/kubernetes#6342]
Today I was trying to help someone debug a mysql service over irc. They didn't have nsenter so kubectl exec didn't work and we just dropped down to using docker exec. Doesn't this defeat the purpose of a unified interface? I feel like there must already be a dicsussion on this but wasn't able to find it.

**Comment by bgrant0607** [comment]
cc @vishh @ncdc

**Comment by vishh** [comment]
#3253 #3763 #5046

To summarize, the docker daemon is not expected to scale to lots of exec
connections. I am not sure if that statement is backed with some
performance testing data.

**Comment by thockin-cc** [comment]
Docker exec will be lower throughput, but might be an acceptable fallback..

**Comment by bprashanth** [comment]
Fwiw I like that option. They were running on bare metal and actually installed nsenter after the first kubectl exec failed, but it didn't work as expeted. After a little dubugging I realized `sudo nsenter` won't work beacuse of secure_path. Recommending changes to /etc/sudoers felt dirty. Maybe the kubelet should use the entire path to nsenter instead? dunno.  

nsenter worked beautifully on my gce minion.

**Comment by vishh** [comment]
As an aside, we should be using nsinit and not nsenter.

**Comment by ncdc** [comment]
@vishh do the GCE node VMs have nsinit installed? I know on Fedora it's in the `golang-github-docker-libcontainer` RPM. I'd have to check to see if that's available for centos and RHEL as well.

One nice thing about `nsenter` (as well as `docker exec`) is that it doesn't rely on the internals of libcontainer's container json files, which, IIRC, moved locations between Docker 1.5 and the upcoming 1.6.

**Comment by vishh** [comment]
I doubt that we are packaging an appropriate version of nsinit with GCE container VM images. That is something we can do. And yes nsinit will need access to the config file. cc @dchen1107 
I have a more fundamental question- Isn't the kubelet likely to be a single point of failure with the current model? 
How about we have the kubelet falling back to using `docker exec` if nsenter/nsinit isn't available or doesn't work?

**Comment by ncdc** [comment]
> Isn't the kubelet likely to be a single point of failure with the current model? 

Yes, but we have plans to address this (#3481)

> How about we have the kubelet falling back to using docker exec if nsenter/nsinit isn't available or doesn't work?

We can do that if everyone agrees. My initial implementation was just to get something working, and we chose nsenter for that.

**Comment by dchen1107** [comment]
ContainerVM does include nsinit, and the current version is 0.1

```
# nsinit --version
nsinit version 0.1
```

**Comment by vishh** [comment]
I tried to parse #3481 and could not infer how the kubelet will disappear from the exec architecture. Can you summarize that? :)
I don't think we have any issues with using `docker exec` directly for now, while we come up with a better long term localhost exec strategy. 
Will `docker exec` work for all the current use cases of `kubectl exec`?

@dchen1107: We might have to package a version of nsinit that is compatible with docker daemon.

**Comment by ncdc** [comment]
@thockin we have some code that uses libcontainer to "docker exec" directly, without needing the Docker daemon or using nsenter. Would you be willing to use that? We'll maintain it (you'd vendor it) and it will be targeted at Kubernetes/OpenShift use cases.

Original libcontainer PR is here: https://github.com/docker/libcontainer/pull/510

/cc @smarterclayton @mrunalp

**Comment by smarterclayton** [comment]
Things we know we need for the Docker container runtime version:

1. Ability to start as a specific user
2. Ability to be run outside the Docker daemon's control
3. Manage security more cleanly

Rocket already comes with this function (right?) so I think it's mostly just about offering parity.

**Comment by smarterclayton** [comment]
It looks like a very small binary that pulls in libcontainer (which we already do) and uses Docker specific paths to lookup the core.

**Comment by ncdc** [comment]
@thockin bump, any comment on https://github.com/GoogleCloudPlatform/kubernetes/issues/6342#issuecomment-97549977? Our repo is here https://github.com/openshift/dockerexec

**Comment by thockin** [comment]
I punt to @vmarmol - as a libcontainer maintainer, he'll be front line on
this when it blows up :)

Still, it seems superior to random half-supported binaries like nsenter

**Comment by thockin** [comment]
I'l double punt to @vmarmol and @vishh

Remind me why docker exec was not viable?  I feel like there was a good
reason but I can't recall it.

**Comment by ncdc** [comment]
@thockin `docker exec` requires all traffic into/out of the container go through the Docker daemon, adding an extra hop and potential bottleneck.

**Comment by thockin** [comment]
Is that *really* a problem?  It feels like that should be the fallback
mode, anyway

**Comment by ncdc** [comment]
I know @smarterclayton wanted to avoid `docker exec` if possible. I know there are some issues with the `nsenter` approach, so it would be nice to use our dockerexec code as option 1, falling back to `docker exec` if necessary.

**Comment by vmarmol** [comment]
@vishh brought up the good point that maintaining our dockerexec will be a cost of its own in the long term (we shouldn't even be using `nsenter` if we want to isolate the user's command). `docker exec` should probably be the default and have the `nsinit`/`nsenter` as a possible route the user can setup.

**Comment by mrunalp** [comment]
@vmarmol @vishh  nsinit doesn't work for docker 1.6+ and so we will need something like dockerexec to replace it unless this PR https://github.com/docker/libcontainer/pull/510 gets merged.

**Comment by vmarmol** [comment]
@mrunalp good point, I think that reinforces the argument for using `docker exec` instead of our own :)

**Comment by vishh** [comment]
Having `docker exec` as the default/fallback option is highly desirable for
providing better user experience. I assume there are no objections to this
approach. We have quite a few open issues around `kubectl exec ` not
working for users.

**Comment by smarterclayton** [comment]
I don't object.  We'll probably only ship Openshift with the direct exec option, and probably aggressively move to improve the security of the execution path (fork a process  into the container that has a one time token, redirect the exec request to that binary) in the long run.

**Comment by smarterclayton** [comment]
You mean, what if I exec in and run "tar cvzf - /mydb" and pipe that back through the daemon's user space goroutine?  :). Or run ten of them?

Most of our exec cases are log streaming, backup, or network based (proxy / port forward) so I just don't want the daemon in the loop at all.  It's fine as a simple default for test/dev setups.

Rocket will have this by default with rkt enter, but that still will go through the kubelet.

**Comment by vishh** [comment]
@dchen1107: Can we prioritize improving exec experience?

**Comment by ncdc** [comment]
I'm planning on making the impl pluggable soon.

**Comment by ncdc** [comment]
I have code that supports using `docker exec` (via go-dockerclient) and I've also got code that works with our openshift/dockerexec implementation. I'd like to make the backend configurable and would like to know the best way to code this. Should we treat the implementations as plugins and allow the administrator to specify which plugin to use when running the kubelet? If not, what approach should we use?

p.s. as a bonus, I've also added SIGWINCH support for terminal window resizing and I'll submit that soon as well :smile:

**Comment by mrunalp** [comment]
Awesome :) I am +1 on making this configurable.

**Comment by smarterclayton** [comment]
I would say make it a kubelet config setting in cmd/kubelet/app.go

----- Original Message -----

**Comment by vishh** [comment]
+1 to making the backend kubelet config. Thanks for working on this Andy!

**Comment by dharmisha** [comment]
I am not able to run the kubectl exec on my kubernetes cluster which is on ubuntu and it just gives an error , upgrade request required. Is there a way to make kubectl exec work?

**Comment by ncdc** [comment]
@dharmisha would you mind asking your question in StackOverflow? This issue was about nsenter vs docker exec.
